### PR TITLE
Icecream support

### PR DESCRIPTION
**Pull request by Emantor**
###### Motivation for this change
Icecream allows distributed compilation in an heterogeneous environment.
This adds the main icecream package and the QT5 based monitor, as well as two new NixOS modules to configure scheduler and daemon. Unfortunately I had to create a new uid for the icecream daemon, since the acceptance of other compile environments hinges on the daemon running as root, which will drop its privileges and switch to the "icecc" user, see [here](https://github.com/icecc/icecream#using-icecream-in-heterogeneous-environments).
I tested this by enabling the module on the desktop and using a debian VM with [ptxdist](https://www.ptxdist.org/), enabling icecc in the options and using [DistroKit](https://git.pengutronix.de/cgit/DistroKit/) as the BSP. A compilation of the armv7a kernel was correctly distributed to both computers running icecream.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/networking/icecream/daemon.nix
Comment (symphorien) [inline nixos/modules/services/networking/icecream/daemon.nix:44]: It does not look like the right description.
Does this path really need to be customizable ? For example I expect the command will break if this contains spaces.

File: nixos/modules/services/networking/icecream/daemon.nix
Comment (symphorien) [inline nixos/modules/services/networking/icecream/daemon.nix:92]: How about setting default=5 then ?

File: nixos/modules/services/networking/icecream/daemon.nix
Comment (symphorien) [inline nixos/modules/services/networking/icecream/daemon.nix:158]:
```suggestion
          cfg.extraConfig
```
The same type of simplification applies above.

File: nixos/modules/services/networking/icecream/scheduler.nix
Comment (symphorien) [inline nixos/modules/services/networking/icecream/scheduler.nix:87]:
```suggestion
          cfg.extraConfig
```
same above

File: nixos/modules/services/networking/icecream/daemon.nix
Comment (Emantor) [reply nixos/modules/services/networking/icecream/daemon.nix:44]: Yes, I need to fix the description. Reasoning behind this was to give the user the ability to put this on a different partition, I'd be fine with removing this too. The compile environments stored there can get quite large, but size can be restricted as well. IDK, open to options.

> Does this path really need to be customizable ? For example I expect the command will break if this contains spaces.

Hm, the `type = types.path` is not enough for that? It needs extra escaping than?

Comment (symphorien) [reply nixos/modules/services/networking/icecream/daemon.nix]: There is no automatic escaping, as escaping rules are not the same for all software.

> Reasoning behind this was to give the user the ability to put this on a different partition

On NixOS it is not much harder to define a bind mount of the partition you want to use on `/var/run/icecc-daemon`.

Anyway, you know the software better than me, so up to you.

**Comment by Emantor** [comment]
I'd like to remove the stale label. I am currently carrying this forward and can rebase onto the master branch quickly, however it is still unclear who needs to review this change.

File: nixos/modules/services/networking/icecream/daemon.nix
Comment (aanderse) [inline nixos/modules/services/networking/icecream/daemon.nix:132]: Already created by `tmpfiles` so this is redundant.

File: nixos/modules/misc/ids.nix
Comment (aanderse) [inline nixos/modules/misc/ids.nix:349]: You shouldn't assign static id for a user unless needed. Use of `isSystemUser` is preferred.

File: nixos/modules/services/networking/icecream/daemon.nix
Comment (aanderse) [inline nixos/modules/services/networking/icecream/daemon.nix:16]: Entirely optional, feel free to ignore, but I'm always partial to `mkEnableOption` .

File: nixos/modules/services/networking/icecream/daemon.nix
Comment (aanderse) [inline nixos/modules/services/networking/icecream/daemon.nix:32]: We have an official policy against opening firewall workout the explicit consent of the user, except for `ssh`. Even in the case that a service is useless without. You could either not provide a default value forcing the user to do so, or add a `warning` if the user hasn't set the value to `true`.

File: nixos/modules/services/networking/icecream/daemon.nix
Comment (aanderse) [inline nixos/modules/services/networking/icecream/daemon.nix:111]: I might call this `extraArgs`, but whatever you feel is more appropriate is fine.

File: nixos/modules/services/networking/icecream/daemon.nix
Comment (aanderse) [inline nixos/modules/services/networking/icecream/daemon.nix:147]: Are you missing a `Group` and `User` declaration here?

File: nixos/modules/services/networking/icecream/daemon.nix
Comment (aanderse) [inline nixos/modules/services/networking/icecream/daemon.nix:112]: Maybe `types.listOf types.str` is a better fit here.

File: nixos/modules/services/networking/icecream/daemon.nix
Comment (aanderse) [inline nixos/modules/services/networking/icecream/daemon.nix:92]: Mentioning default seems a little redundant because you have already specified the `default` which will be presented to the user.

File: nixos/modules/services/networking/icecream/daemon.nix
Comment (aanderse) [inline nixos/modules/services/networking/icecream/daemon.nix:114]: Missing a trailing period.

File: nixos/modules/services/networking/icecream/daemon.nix
Comment (aanderse) [inline nixos/modules/services/networking/icecream/daemon.nix:151]: Maybe you should split these `optionalString` entries out of the list and use `optional`/`optionals` with list concatenation instead.

File: nixos/modules/services/networking/icecream/scheduler.nix
Comment (aanderse) [inline nixos/modules/services/networking/icecream/scheduler.nix:14]: See `mkEnableOption` comment above.

File: nixos/modules/services/networking/icecream/scheduler.nix
Comment (aanderse) [inline nixos/modules/services/networking/icecream/scheduler.nix:26]: You're missing a number of trailing periods in this module.

File: nixos/modules/services/networking/icecream/scheduler.nix
Comment (aanderse) [inline nixos/modules/services/networking/icecream/scheduler.nix:31]: `types.port`

File: nixos/modules/services/networking/icecream/scheduler.nix
Comment (aanderse) [inline nixos/modules/services/networking/icecream/scheduler.nix:62]: See comments from above.

File: nixos/modules/services/networking/icecream/scheduler.nix
Comment (aanderse) [inline nixos/modules/services/networking/icecream/scheduler.nix:85]: See comment above about `optional`/`optionals`.

File: pkgs/applications/networking/icemon/default.nix
Comment (aanderse) [inline pkgs/applications/networking/icemon/default.nix:5]: "3.3"

File: pkgs/applications/networking/icemon/default.nix
Comment (aanderse) [inline pkgs/applications/networking/icemon/default.nix:10]: `rev = "v${version}";`

File: pkgs/applications/networking/icemon/default.nix
Comment (aanderse) [inline pkgs/applications/networking/icemon/default.nix:14]: Already implied by usage of `cmake`.

File: pkgs/applications/networking/icemon/default.nix
Comment (aanderse) [inline pkgs/applications/networking/icemon/default.nix:19]: `maintainer`?

File: pkgs/servers/icecream/default.nix
Comment (aanderse) [inline pkgs/servers/icecream/default.nix:18]: `maintainer`?

**Review by aanderse — CHANGES_REQUESTED** [review]
This PR looks fine to me. There are a few minor issues, but aside from that I recommend rebase and merge. I'll be happy to do so once the PR is in a ready state.

Thanks for your hard work, and your patience. Sometimes things just slip by. Sometimes it can help if you make noise on IRC or discourse.

**Review by aanderse — APPROVED** [review]
Looks great! Does anyone have any reason we shouldn't merge this? It looks like we should.

**Review by infinisil — CHANGES_REQUESTED** [review]

**Comment by Emantor** [comment]
Thanks for the review @Infinisil, I have force pushed another version which should address the review comments.

File: nixos/modules/services/networking/icecream/scheduler.nix
Comment (Emantor) [reply nixos/modules/services/networking/icecream/scheduler.nix]: This should only be open if `openTelnet` is set, adjusted in the new version.

File: nixos/modules/services/networking/icecream/daemon.nix
Comment (Emantor) [reply nixos/modules/services/networking/icecream/daemon.nix]: I dug into the source code and found out that the `libcap-ng` support removes the requirement to start as root to receive remote compile environments documented in the README. The service now uses `DynamicUser` and `AmbientCapabilities` with `RuntimeDirectory` and `StateDirectory` and works fine with YOCTO dunfell and an OpenSSL test compilation. The basedir and user creation has been removed in the process.

File: nixos/modules/services/networking/icecream/daemon.nix
Comment (Emantor) [reply nixos/modules/services/networking/icecream/daemon.nix]: I did not know about `escapeShellArgs`, adjusted in the new version.

**Comment by Emantor** [comment]
@Infinisil gentle ping

File: nixos/modules/services/networking/icecream/daemon.nix
Comment (Infinisil) [inline nixos/modules/services/networking/icecream/daemon.nix:40]: What does this `null` represent? This isn't obvious from the option description. I suggest adding an explanation for each of the `null`able options.

Comment (Emantor) [reply nixos/modules/services/networking/icecream/daemon.nix]: Done.

**Comment by Emantor** [comment]
@Infinisil gentle ping

**Comment by Emantor** [comment]
@Infinisil gentle ping

**Review by infinisil — APPROVED** [review]
Sorry for the delay, looks good now!

**Pull request merged**
